src/openai/_tracing.py

```diff
@@ -0,0 +1,160 @@
+    _set_attribute(span, "gen_ai.request.top_p", kwargs.get("top_p"))
+
+
+def _add_response_chat_message_event(span: trace.Span, result: ChatCompletion) -> None:
+    for choice in result.choices:
+        response = {
+            "message.role": choice.message.role,
+            "message.content": choice.message.content,
+            "finish_reason": choice.finish_reason,
+            "index": choice.index,
+        }
+        if choice.message.tool_calls:
+            response.update({"message.tool_calls": "".join([tool.to_json() for tool in choice.message.tool_calls])})
+        span.add_event(name="gen_ai.response.message", attributes={"event.data": json.dumps(response, indent=2)})
+
+
+def _add_response_chat_attributes(span: trace.Span, result: ChatCompletion) -> None:
+    _set_attribute(span, "gen_ai.response.id", result.id)
+    _set_attribute(span, "gen_ai.response.model", result.model)
+    _set_attribute(span, "gen_ai.response.finish_reason", result.choices[0].finish_reason)
+    if hasattr(result, "usage"):
+        _set_attribute(span, "gen_ai.usage.completion_tokens", result.usage.completion_tokens if result.usage else None)
+        _set_attribute(span, "gen_ai.usage.prompt_tokens", result.usage.prompt_tokens if result.usage else None)
+
+
+def _traceable_stream(stream_obj: Stream[ChatCompletionChunk], span: trace.Span) -> Generator[ChatCompletionChunk, None, None]:
+    try:
+        accumulate: dict[str, Any] = {"role": ""}
+        for chunk in stream_obj:
+            for item in chunk.choices:
+                if item.finish_reason:
+                    accumulate["finish_reason"] = item.finish_reason
+                if item.index:
+                    accumulate["index"] = item.index
+                if item.delta.role:
+                    accumulate["role"] = item.delta.role
+                if item.delta.content:
+                    accumulate.setdefault("content", "")
+                    accumulate["content"] += item.delta.content
+                if item.delta.tool_calls:
+                    accumulate.setdefault("tool_calls", [])
+                    for tool_call in item.delta.tool_calls:
+                        if tool_call.id:
+                            accumulate["tool_calls"].append({"id": tool_call.id, "type": "", "function": {"name": "", "arguments": ""}})
+                        if tool_call.type:
+                            accumulate["tool_calls"][-1]["type"] = tool_call.type
+                        if tool_call.function and tool_call.function.name:
+                            accumulate["tool_calls"][-1]["function"]["name"] = tool_call.function.name
+                        if tool_call.function and tool_call.function.arguments:
+                            accumulate["tool_calls"][-1]["function"]["arguments"] += tool_call.function.arguments
+            yield chunk
+
+        span.add_event(name="gen_ai.response.message", attributes={"event.data": json.dumps(accumulate, indent=2)})
```

Comment:
Nit: we can save space by not indenting things in the json payload. Not sure if we should expect tools to present the contents in a nice way rather than bloating the value with pretty-print spaces... but I suspect that will be what people want to production workloads...